precompile/allowlist/test_allowlist.go

```diff
@@ -0,0 +1,271 @@
+			Caller: TestAdminAddr,
+			InputFn: func(t *testing.T) []byte {
+				input, err := PackModifyAllowList(TestEnabledAddr, NoRole)
+				require.NoError(t, err)
+
+				return input
+			},
+			SuppliedGas: ModifyAllowListGasCost - 1,
+			ReadOnly:    false,
+			ExpectedErr: vmerrs.ErrOutOfGas.Error(),
+		},
+		"read allow list no role": {
+			Caller:      TestNoRoleAddr,
+			Input:       PackReadAllowList(TestNoRoleAddr),
+			SuppliedGas: ReadAllowListGasCost,
+			ReadOnly:    false,
+			ExpectedRes: common.Hash(NoRole).Bytes(),
+		},
+		"read allow list admin role": {
+			Caller:      TestAdminAddr,
+			Input:       PackReadAllowList(TestAdminAddr),
+			SuppliedGas: ReadAllowListGasCost,
+			ReadOnly:    false,
+			ExpectedRes: common.Hash(AdminRole).Bytes(),
+		},
+		"read allow list with readOnly enabled": {
+			Caller:      TestAdminAddr,
+			Input:       PackReadAllowList(TestNoRoleAddr),
+			SuppliedGas: ReadAllowListGasCost,
+			ReadOnly:    true,
+			ExpectedRes: common.Hash(NoRole).Bytes(),
+		},
+		"read allow list out of gas": {
+			Caller:      TestAdminAddr,
+			Input:       PackReadAllowList(TestNoRoleAddr),
+			SuppliedGas: ReadAllowListGasCost - 1,
+			ReadOnly:    true,
+			ExpectedErr: vmerrs.ErrOutOfGas.Error(),
+		},
+		"initial config sets admins": {
+			Config: mkConfigWithAllowList(
+				module,
+				&AllowListConfig{
+					AdminAddresses: []common.Address{TestNoRoleAddr, TestEnabledAddr},
+				},
+			),
+			SuppliedGas: 0,
+			ReadOnly:    false,
+			AfterHook: func(t *testing.T, state contract.StateDB) {
+				require.Equal(t, AdminRole, GetAllowListStatus(state, contractAddress, TestNoRoleAddr))
+				require.Equal(t, AdminRole, GetAllowListStatus(state, contractAddress, TestEnabledAddr))
+			},
+		},
+		"initial config sets enabled": {
+			Config: mkConfigWithAllowList(
+				module,
+				&AllowListConfig{
+					EnabledAddresses: []common.Address{TestNoRoleAddr, TestAdminAddr},
+				},
+			),
+			SuppliedGas: 0,
+			ReadOnly:    false,
+			AfterHook: func(t *testing.T, state contract.StateDB) {
+				require.Equal(t, EnabledRole, GetAllowListStatus(state, contractAddress, TestAdminAddr))
+				require.Equal(t, EnabledRole, GetAllowListStatus(state, contractAddress, TestNoRoleAddr))
+			},
+		},
+	}
+}
+
+func RunPrecompileWithAllowListTests(t *testing.T, module modules.Module, newStateDB func(t *testing.T) contract.StateDB, contractTests map[string]testutils.PrecompileTest) {
+	tests := AllowListTests(module)
+	// Add the contract specific tests to the map of tests to run.
+	for name, test := range contractTests {
+		tests[name] = test
+	}
+
+	contractAddress := module.Address
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			state := newStateDB(t)
+
+			// Set up the state so that each address has the expected permissions at the start.
+			SetAllowListRole(state, contractAddress, TestAdminAddr, AdminRole)
+			SetAllowListRole(state, contractAddress, TestEnabledAddr, EnabledRole)
```

Comment:
Can we change this to a default `BeforeHook` ? I think this could overlap with the behavior in initial config